storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -122,33 +122,41 @@ public ConsumerTask(RemotePartitionMetadataEventHandler remotePartitionMetadataE
     public void run() {
         log.info("Starting consumer task thread.");
         while (!isClosed) {
-            try {
-                if (hasAssignmentChanged) {
-                    maybeWaitForPartitionAssignments();
-                }
+            ingestRecords();
+        }
+        log.info("Exited from consumer task thread");
+    }
 
-                log.trace("Polling consumer to receive remote log metadata topic records");
-                final ConsumerRecords<byte[], byte[]> consumerRecords = consumer.poll(Duration.ofMillis(pollTimeoutMs));
-                for (ConsumerRecord<byte[], byte[]> record : consumerRecords) {
-                    processConsumerRecord(record);
-                }
-                maybeMarkUserPartitionsAsReady();
-            } catch (final WakeupException ex) {
-                // ignore logging the error
-                isClosed = true;
-            } catch (final RetriableException ex) {
-                log.warn("Retriable error occurred while processing the records. Retrying...", ex);
-            } catch (final Exception ex) {
-                isClosed = true;
-                log.error("Error occurred while processing the records", ex);
+    public void ingestRecords() {
+        try {
+            if (hasAssignmentChanged) {
+                maybeWaitForPartitionAssignments();
+            }
+
+            log.trace("Polling consumer to receive remote log metadata topic records");
+            final ConsumerRecords<byte[], byte[]> consumerRecords = consumer.poll(Duration.ofMillis(pollTimeoutMs));
+            for (ConsumerRecord<byte[], byte[]> record : consumerRecords) {
+                processConsumerRecord(record);
             }
+            maybeMarkUserPartitionsAsReady();
+        } catch (final WakeupException ex) {
+            // ignore logging the error
+            closeConsumer();
+        } catch (final RetriableException ex) {
+            log.warn("Retriable error occurred while processing the records. Retrying...", ex);
+        } catch (final Exception ex) {
+            log.error("Error occurred while processing the records", ex);
+            closeConsumer();
         }
+    }
+
+    private void closeConsumer() {
```

Comment:
This introduces a race condition that can cause a consumer leak. Consider the following sequence:

1. The ConsumerTask thread returns from maybeMarkPartitionsAsReady(), and then pauses
2. An external thread calls close(), and sets isClose = true
3. The ConsumerTask thread resumes, returns from ingestRecords(), and checks the isClose flag in run().
4. The ConsumerTask thread returns from run() and exits, without ever closing the consumer.